pkg/secretsencrypt/controller.go

```diff
@@ -0,0 +1,207 @@
+package secretsencrypt
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/rancher/k3s/pkg/cluster"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/generated/clientset/versioned/scheme"
+	coreclient "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"k8s.io/client-go/kubernetes"
+	coregetter "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/pager"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	controllerAgentName        string = "reencrypt-controller"
+	secretsUpdateStartEvent    string = "SecretsUpdateStart"
+	secretsProgressEvent       string = "SecretsProgress"
+	secretsUpdateCompleteEvent string = "SecretsUpdateComplete"
+	secretsUpdateErrorEvent    string = "SecretsUpdateError"
+)
+
+type handler struct {
+	ctx           context.Context
+	controlConfig *config.Control
+	nodes         coreclient.NodeController
+	secrets       coreclient.SecretController
+	recorder      record.EventRecorder
+}
+
+func Register(
+	ctx context.Context,
+	kubernetes kubernetes.Interface,
+	controlConfig *config.Control,
+	nodes coreclient.NodeController,
+	secrets coreclient.SecretController,
+	events coregetter.EventInterface,
+) error {
+	h := &handler{
+		ctx:           ctx,
+		controlConfig: controlConfig,
+		nodes:         nodes,
+		secrets:       secrets,
+		recorder:      buildEventRecorder(events),
+	}
+
+	nodes.OnChange(ctx, "reencrypt-controller", h.onChangeNode)
+	return nil
+}
+
+func buildEventRecorder(events coregetter.EventInterface) record.EventRecorder {
+	// Create event broadcaster
+	logrus.Info("Creating reencrypt event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	eventBroadcaster.StartRecordingToSink(&coregetter.EventSinkImpl{Interface: events})
+	return eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+}
```

Comment:
this should probably be exported from the util package; there are other places we could reuse this such as https://github.com/k3s-io/k3s/blob/05d43278e29cd7b2cb426667aa1e24672cee5765/pkg/deploy/controller.go#L77-L81

Also, it'd probably be good to use `schemes.All` and set the `Host` field in the recorder's EventSource, as we do in that existing code.

Reply:
Consolidated into single util function.